R/corPFSOS.R

```diff
@@ -0,0 +1,307 @@
+#'
+#' @return This returns the value of PFS*OS survival function at time t.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' survPFSOS(0.4, transition)
+survPFSOS <- function(t, transition) {
+  sapply(t, function(x) {
+    intval <- stats::integrate(PFSOSInteg, lower = 0, upper = sqrt(x), x, transition)$value
+    survPFS(transition, sqrt(x)) + intval
+  })
+}
+
+#' Correlation of PFS and OS event times for Different Transition Models
+#'
+#' @param transition (`TransitionParameters`)\cr
+#'   see [exponential_transition()], [weibull_transition()] or [piecewise_exponential()] for details.
+#'
+#' @return The correlation of PFS and OS.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' corTrans(transition)
+corTrans <- function(transition) {
+  # E(PFS) & E(OS).
+  expvalPFS <- stats::integrate(survPFS,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  expvalOS <- stats::integrate(survOS,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  # Var(PFS) & Var(OS).
+  expvalPFS2 <- 2 * stats::integrate(expvalPFSInteg,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  expvalOS2 <- 2 * stats::integrate(expvalOSInteg,
+    lower = 0, upper = Inf,
+    transition = transition
+  )$value
+
+  varPFS <- expvalPFS2 - expvalPFS^2
+
+  varOS <- expvalOS2 - expvalOS^2
+
+  # E(PFS*OS).
+  expvalPFSOS <- stats::integrate(survPFSOS,
+    lower = 0, upper = Inf,
+    transition
+  )$value
+
+  # Cor(PFS, OS).
+  (expvalPFSOS - expvalPFS * expvalOS) / (varPFS * varOS)^0.5
+}
+
+#' Correlation of PFS and OS event times for data from the IDM
+#'
+#' @param data (`data.frame`)\cr in the format produced by [getOneClinicalTrial()].
+#' @param transition (`TransitionParameters` object)\cr specifying the assumed distribution of transition hazards.
+#'   Initial parameters for optimization can be specified here.
+#'   See [exponential_transition()] or [weibull_transition()] for details.
+#' @param bootstrap (`logical`)\cr if `TRUE` computes confidence interval via bootstrap.
+#' @param bootstrap.n (`integer`)\cr number of bootstrap samples.
+#' @param bootstrap.width (`positive `number`)\cr confidence interval for the correlation estimate.
+#'
+#' @return The correlation of PFS and OS.
+#' @export
+#'
+#' @examples
+#' transition <- exponential_transition(h01 = 1.2, h02 = 1.5, h12 = 1.6)
+#' data <- getClinicalTrials(
+#'   nRep = 10, nPat = c(50), seed = 1234, datType = "1rowTransition",
+#'   transitionByArm = list(transition), dropout = list(rate = 0.5, time = 12),
+#'   accrual = list(param = "intensity", value = 7)
+#' )[[1]]
+#' corPFSOS(data, transition)
+corPFSOS <- function(data, transition, bootstrap = TRUE, bootstrap.n = 100, bootstrap.width = 0.95) {
+  if (bootstrap == TRUE) {
```

Comment:
```suggestion
  if (bootstrap) {
```
because after assertion this is safe to do